File: message/message.py
import struct
from typing import Any, Dict


class Message:

    def __init__(self):
        self.options: Dict[str, Any] = {}

    def add_option(self, key: str, value: Any):
        self.options[key] = value
        return self

    def add_options(self, options: Dict[str, Any]):
        for key, value in options.items():
            self.options[key] = value

        return self
    
    def remove_options(self, *keys: str):
        for key in keys:
            if key in self.options:
                del self.options[key]

        return self
    
    def option(self, key: str):
        return self.options.get(key, None)
    
    def option_as_str(self, key: str):
        value = self.options.get(key, None)
        if isinstance(value, str):
            return value
        elif isinstance(value, (bytes, bytearray)):
            return value.decode('utf-8')
        else:
            return None
    
    def option_as_int(self, key: str):
        value = self.options.get(key, None)
        if isinstance(value, int):
            return value
        elif isinstance(value, (bytes, bytearray)):
            return int(value.decode('utf-8'))
        else:
            return None

    def option_as_byte(self, key: str):
        return self.option_as_int(key)
    
    def option_as_bytes(self, key: str):
        value = self.options.get(key, None)
        if isinstance(value, (bytes, bytearray)):
            return value
        elif value is None:
            return None
        else:
            return str(value).encode('utf-8')
    
    def get_options(self):
        return self.options
    
    def clear(self):
        self.options.clear()
        return self
    
    def count(self) -> int:
        return len(self.options)
    
    def copy(self):
        return Message().add_options(self.options)

File: app/src/main/java/capstone/design/message/Message.java
package capstone.design.message;

import java.nio.charset.StandardCharsets;
import java.util.HashMap;
import java.util.Map;
import capstone.design.Utils;

public class Message {

    private final Map<String, Object> options = new HashMap<>();

    private Message() {}

    public static Message of() { return new Message(); }
    public static Message of(String key, Object value) { return new Message().addOption(key, value); }
    public static Message of(Map<String, Object> options) { return new Message().addOptions(options); }
    
    public Message addOption(String key, Object value) {
        Utils.validate(key, value);

        options.put(key, value);
        return this;
    }

    public Message addOptions(Map<String, Object> options) {
        Utils.validate(options);
        
        options.forEach(this::addOption);
        return this;
    }

    public Message removeOptions(String... keys) {
        for (String key : keys) {
            options.remove(key);
        }

        return this;
    }

    public Object option(String key) {
        return options.get(key);
    }

    public Integer optionAsInt(String key) {
        switch (options.get(key)) {
            case Integer i -> { return i; }
            case byte[] b -> { return Integer.parseInt(new String(b, StandardCharsets.UTF_8)); }
            case null, default -> { return null; }
        }
    }

    public Long optionAsLong(String key) {
        switch (options.get(key)) {
            case Long l -> { return l; }
            case byte[] b -> { return Long.parseLong(new String(b, StandardCharsets.UTF_8)); }
            case null, default -> { return null; }
        }
    }

    public String optionAsString(String key) {
        switch (options.get(key)) {
            case String s -> { return s; }
            case byte[] b -> { return new String(b, StandardCharsets.UTF_8); }
            case null, default -> { return null; }
        }
    }

    public Byte optionAsByte(String key) {
        switch (options.get(key)) {
            case Byte b -> { return b; }
            case byte[] b -> { return Byte.parseByte(new String(b, StandardCharsets.UTF_8)); }
            case null, default -> { return null; }
        }
    }

    public byte[] optionAsBytes(String key) {
        switch (options.get(key)) {
            case byte[] b -> { return b; }
            case null, default -> { return null; }
        }
    }

    public Map<String, Object> options() {
        return options;
    }

    public Message clear() {
        options.clear();
        
        return this;
    }

    public Message copy() {
        return Message.of(options);
    }

    @Override
    public String toString() {
        String str = "";
        for (Map.Entry<String, Object> entry : options.entrySet()) {
            str += entry.getKey() + ": ";
            switch (entry.getValue()) {
                case byte[] b -> str += new String(b, StandardCharsets.UTF_8) + ", ";
                default -> str += entry.getValue() + ", ";
            }
        }

        return "Message{" + str + "}";
    }
}
